### ISSUE TITLE
s_client gets fried by ENGINE_by_id("openssl") in linked library

### ISSUE THREAD

**Issue by AndreasFuchsTPM** [openssl/openssl#8490]
Since I'm clueless otherwise, I was wondering if this was an upstream issue ?

using `openssl s_client` a user of the tpm2-tss engine gets errors related to ENGINE_by_id attempting to load openssl.so ? https://github.com/tpm2-software/tpm2-tss-engine/issues/93
Which I could also reproduce with "OpenSSL 1.1.0j  20 Nov 2018"
My code does not include any call to ENGINE_by_id at all.

Is that issue known ?
Note that s_client continues nicely afterwards and performs the TLS connection correctly.

```
139897488847296:error:25066067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:113:filename(~/tpm2-tss-engine/.libs/openssl.so): ~/tpm2-tss-engine/.libs/openssl.so: cannot open shared object file: No such file or directory
139897488847296:error:25070067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:161:
139897488847296:error:260B6084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:414:
139897488847296:error:2606A074:engine routines:ENGINE_by_id:no such engine:crypto/engine/eng_list.c:339:id=openssl
139897488847296:error:25066067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:113:filename(~/tpm2-tss-engine/.libs/openssl.so): ~/tpm2-tss-engine/.libs/openssl.so: cannot open shared object file: No such file or directory
139897488847296:error:25070067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:161:
139897488847296:error:260B6084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:414:
139897488847296:error:2606A074:engine routines:ENGINE_by_id:no such engine:crypto/engine/eng_list.c:339:id=openssl
```

**Comment by levitte** [comment]
It's the ~. The shell is supposed to expand that into a proper path

**Comment by AndreasFuchsSIT** [comment]
Yes, sorry. The ~ was added by me for reasons of anonymization... Should've known... :-(

My question is basically, what is the engine "openssl.so" and why is it search for. I could not find any reference anywhere...

**Comment by levitte** [comment]
Hmmm, there is such a thing, but it's a test engine that should not end up in production or be loaded automatically.

Have you looked in the default configuration file (i.e openssl.cnf)?

**Comment by AndreasFuchsSIT** [comment]
```
$ echo $OPENSSL_CONF
/home/afuchs/Dokumente/oss-tss/INSTALLDIR/usr/bin/../../usr/openssl/openssl.cnf
```
which is attached (added .txt for github)
[openssl.cnf.txt](https://github.com/openssl/openssl/files/2971148/openssl.cnf.txt)

**Comment by levitte** [comment]
Ok...  I tried the first s_client command I saw in  tpm2-software/tpm2-tss-engine#93, and here's what I got:

```
: ; openssl s_client -connect somewebsite.com:443 -cert rsa.crt -key rsa.tss -CAfile CA.pem -engine tpm2tss -keyform engine
invalid engine "tpm2tss"
140179301053504:error:25066067:DSO support routines:dlfcn_load:could not load the shared library:../crypto/dso/dso_dlfcn.c:119:filename(/usr/lib/x86_64-linux-gnu/engines-1.1/tpm2tss.so): /usr/lib/x86_64-linux-gnu/engines-1.1/tpm2tss.so: cannot open shared object file: No such file or directory
140179301053504:error:25070067:DSO support routines:DSO_load:could not load the shared library:../crypto/dso/dso_lib.c:162:
140179301053504:error:260B6084:engine routines:dynamic_load:dso not found:../crypto/engine/eng_dyn.c:414:
140179301053504:error:2606A074:engine routines:ENGINE_by_id:no such engine:../crypto/engine/eng_list.c:334:id=tpm2tss
140179301053504:error:25066067:DSO support routines:dlfcn_load:could not load the shared library:../crypto/dso/dso_dlfcn.c:119:filename(libtpm2tss.so): libtpm2tss.so: cannot open shared object file: No such file or directory
140179301053504:error:25070067:DSO support routines:DSO_load:could not load the shared library:../crypto/dso/dso_lib.c:162:
140179301053504:error:260B6084:engine routines:dynamic_load:dso not found:../crypto/engine/eng_dyn.c:414:
no engine specified
unable to load client certificate private key file
```

It's the exact same kind of output, but make more sense

**Comment by levitte** [comment]
I think someone will have to run a debug session to see the call stack each time `ENGINE_by_id` is called.  Would you mind?

**Comment by AndreasFuchsSIT** [comment]
Ok, I found it... It's actually in
https://github.com/tpm2-software/tpm2-tss/blob/1d57398857b6302acbec7a601b49cb46745a827e/src/tss2-esys/esys_crypto_ossl.c#L33

Funny enough, never shows up in the integration tests of tpm2-tss...

Thanks a lot the help.
Sorry for the confusion.

**Comment by AndreasFuchsSIT** [comment]
@levitte Maybe one last question, if I may:
How do you ask OpenSSL for an engine that comes from software ?
I guess, the offending code was relying on
https://github.com/openssl/openssl/blob/52eb5b7ca46a1bf0bb0e148c651eb7dcaae1a074/crypto/engine/eng_openssl.c#L77

What should be used here instead, if we want to force a software implementation of ciphers ?
https://github.com/tpm2-software/tpm2-tss/blob/1d57398857b6302acbec7a601b49cb46745a827e/src/tss2-esys/esys_crypto_ossl.c#L33

**Comment by levitte** [comment]
> Sorry for the confusion.

No problem.

I'm not sure what you mean with forcing a software implementation of ciphers. If you're calling things like `EVP_` functions, you've linked with libcrypto, which contains all the OpenSSL software implementations.  The way I see it, all the `get_engine()` calls there can be replaced with `NULL`...  or that's at least what it looks to me after just a quick look.

**Comment by AndreasFuchsSIT** [comment]
@levitte One more question: What happens, if the default OpenSSL engine is set to tpm2tss ?
i.e. openssl -> tpm2-tss-engine -> tpm2-tss -> openssl -> tpm2-tss-engine, ...

I assume the ENGINE_by_id was added to prevent these kind of cycle invocations.

Could this happen or does NULL enforce the software-implementation or how should we do this ?

Thanks a lot for the help !

**Comment by AndreasFuchsSIT** [comment]
Also, now that I think about it: If ENGINE_by_id("openssl") errors, it will return NULL. So apparently tpm2-tss will have used NULL all along.
However, changing it to NULL without calling ENGINE_by_id, the bug in the tpm2-tss-engine's s_client test suddenly disappears.

Seems like a failure in ENGINE_by_id is not side-effect free wrt s_client ?

**Comment by AndreasFuchsSIT** [comment]
I'm reopening this, since IMHO the side-effect of ENGINE_by_id from a linked library into the s_client app is not only weird but probably broken ?

**Comment by levitte** [comment]
I'm not sure I understand the issue... What are you trying to do?

**Comment by AndreasFuchsSIT** [comment]
So `openssl s_client` behaves differently when used with `-engine tpm2tss` from without. The difference is that with `tpm2tss` it exits after successful session establishment opposed to remaining.

When I exchange `ENGINE_by_id("openssl")` which returns `NULL` with a plain `NULL` inside the tpm2-tss library, then `openssl s_client -engine tpm2tss` behaves the same as without the engine.

Thus, the failing call to `ENGINE_by_id("openssl")` does have some side-effect (not along the call path since it returns `NULL`) that makes `s_client` exit _after_ successful connection establishment.

P.S. s_server works just fine under the same circumstances though...

**Comment by AndreasFuchsSIT** [comment]
Just to make clear what this means in praxis:
This test https://github.com/tpm2-software/tpm2-tss-engine/blob/master/test/sclient.sh only works if this commit https://github.com/tpm2-software/tpm2-tss/commit/4bd982186d5d408f92e329660d85f8c0649cb6f8 was applied.

**Comment by nhorman** [comment]
Is this problem still occuring on the 3.3 release?

**Comment by nhorman** [comment]
marking as inactive, for lack of response, to be closed at the end of 3.4 dev
